<h1>Project: UK Road Accidents</h1>

<h2>Analyst:</h2>

<h2>Zairesh Castillo</h2>

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('accident_data.csv')

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [5]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [6]:
accident["Index"]

0         200701BS64157
1         200701BS65737
2         200701BS66127
3         200701BS66128
4         200701BS66837
              ...      
660674    201091NM01760
660675    201091NM01881
660676    201091NM01935
660677    201091NM01964
660678    201091NM02142
Name: Index, Length: 660679, dtype: object

In [7]:
accident["Accident_Severity"]

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [8]:
accident["Accident Date"]

0         05/06/2019
1         02/07/2019
2         26/08/2019
3         16/08/2019
4         03/09/2019
             ...    
660674    18/02/2022
660675    21/02/2022
660676    23/02/2022
660677    23/02/2022
660678    28/02/2022
Name: Accident Date, Length: 660679, dtype: object

In [9]:
accident["Latitude"]

0         51.506187
1         51.495029
2         51.517715
3         51.495478
4         51.488576
            ...    
660674    57.374005
660675    57.232273
660676    57.585044
660677    57.214898
660678    57.575210
Name: Latitude, Length: 660679, dtype: float64

In [10]:
accident["Light_Conditions"]

0               Darkness - lights lit
1                            Daylight
2         Darkness - lighting unknown
3                            Daylight
4               Darkness - lights lit
                     ...             
660674                       Daylight
660675         Darkness - no lighting
660676                       Daylight
660677         Darkness - no lighting
660678                       Daylight
Name: Light_Conditions, Length: 660679, dtype: object

In [11]:
accident["District Area"]

0         Kensington and Chelsea
1         Kensington and Chelsea
2         Kensington and Chelsea
3         Kensington and Chelsea
4         Kensington and Chelsea
                   ...          
660674                  Highland
660675                  Highland
660676                  Highland
660677                  Highland
660678                  Highland
Name: District Area, Length: 660679, dtype: object

In [12]:
accident["Longitude"]

0        -0.209082
1        -0.173647
2        -0.210215
3        -0.202731
4        -0.192487
            ...   
660674   -3.467828
660675   -3.809281
660676   -3.862727
660677   -3.823997
660678   -3.895673
Name: Longitude, Length: 660679, dtype: float64

In [13]:
accident["Number_of_Casualties"]

0         1
1         1
2         1
3         1
4         1
         ..
660674    2
660675    1
660676    1
660677    1
660678    1
Name: Number_of_Casualties, Length: 660679, dtype: int64

In [14]:
accident["Number_of_Vehicles"]

0         2
1         2
2         3
3         4
4         2
         ..
660674    1
660675    1
660676    3
660677    2
660678    1
Name: Number_of_Vehicles, Length: 660679, dtype: int64

In [15]:
accident["Road_Surface_Conditions"]

0                  Dry
1          Wet or damp
2                  Dry
3                  Dry
4                  Dry
              ...     
660674             Dry
660675    Frost or ice
660676    Frost or ice
660677     Wet or damp
660678     Wet or damp
Name: Road_Surface_Conditions, Length: 660679, dtype: object

In [16]:
accident["Road_Type"]

0         Single carriageway
1         Single carriageway
2                        NaN
3         Single carriageway
4                        NaN
                 ...        
660674    Single carriageway
660675    Single carriageway
660676    Single carriageway
660677    Single carriageway
660678      Dual carriageway
Name: Road_Type, Length: 660679, dtype: object

In [17]:
accident["Urban_or_Rural_Area"]

0         Urban
1         Urban
2         Urban
3         Urban
4         Urban
          ...  
660674    Rural
660675    Rural
660676    Rural
660677    Rural
660678    Rural
Name: Urban_or_Rural_Area, Length: 660679, dtype: object

In [18]:
accident["Weather_Conditions"]

0            Fine no high winds
1         Raining no high winds
2                           NaN
3            Fine no high winds
4                           NaN
                  ...          
660674       Fine no high winds
660675       Fine no high winds
660676       Fine no high winds
660677       Fine no high winds
660678    Snowing no high winds
Name: Weather_Conditions, Length: 660679, dtype: object

In [19]:
accident["Vehicle_Type"]

0                                          Car
1                                          Car
2                        Taxi/Private hire car
3         Bus or coach (17 or more pass seats)
4                                Other vehicle
                          ...                 
660674                                     Car
660675                                     Car
660676                                     Car
660677                   Motorcycle over 500cc
660678                                     Car
Name: Vehicle_Type, Length: 660679, dtype: object

In [20]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [21]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Road_Type'] = accident['Road_Type'].fillna(accident['Road_Type'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna(accident['Weather_Conditions'].mode()[0])

In [22]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [23]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [24]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [25]:
accident.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

In [26]:
accident['year'] = accident['Accident Date'].dt.year
accident['month'] = accident['Accident Date'].dt.month
accident['day'] = accident['Accident Date'].dt.day
accident['dayofweek'] = accident['Accident Date'].dt.dayofweek

In [27]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
year                       0
month                      0
day                        0
dayofweek                  0
dtype: int64

<h1>QUESTIONS</h1>

<h2>1. how many accidents fall into each category of accident_severinity?</h2>

In [28]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h3>insigth: there are 563801 who's Slightly enjured, 88217 who's seriopusly injured and 8661 who's fatal. </h3>

<h1>2. what is the average number of casualties for each severity level?</h1>

In [29]:
accident.groupby('Accident_Severity')['Number_of_Casualties'].mean()

Accident_Severity
Fatal      1.903129
Serious    1.467280
Slight     1.331402
Name: Number_of_Casualties, dtype: float64

<h2>insigth: the number of casualties in Fatal is 1.903129, 1.467280 is serious and 1.331402 is sligth. </h2>

<h1>3. which District Area has the highest number of fatal accidents?</h1>

In [30]:
accident[accident['Accident_Severity'] == 'Fatal']['District Area'].value_counts()

District Area
Birmingham                  105
Leeds                        93
Highland                     88
East Riding of Yorkshire     85
Bradford                     71
                           ... 
Havant                        2
Blyth Valley                  2
Clackmannanshire              1
Harlow                        1
Stevenage                     1
Name: count, Length: 422, dtype: int64

<h2>insigth: area that has the the highest number of fatal accidents is Birmingham with the number of 105</h2>

<h1> 4.which vehicle type is the most often involve in fatal accident? </h1>

In [31]:
accident[accident['Accident_Severity'] == 'Fatal']['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      6577
Van / Goods 3.5 tonnes mgw or under       467
Motorcycle over 500cc                     339
Bus or coach (17 or more pass seats)      325
Goods 7.5 tonnes mgw and over             216
Motorcycle 125cc and under                189
Taxi/Private hire car                     155
Motorcycle over 125cc and up to 500cc     105
Motorcycle 50cc and under                  95
Other vehicle                              70
Goods over 3.5t. and under 7.5t            67
Minibus (8 - 16 passenger seats)           29
Agricultural vehicle                       21
Pedal cycle                                 6
Name: count, dtype: int64

<h2>insigth: car is the vehicle that is often involve in a fatal accident.</h2>

<h1> 5. on average, how many vehicles are involve in accidents with serious severity?</h1>

In [32]:
accident[accident['Accident_Severity'] == 'Serious']['Number_of_Vehicles'].mean()

1.6783273065282203

<h2>insigth: the average number of vehicles that is involve in an serious severity accident are 1.6783273065282203. </h2>

<h1> 6. which month records the highest number of accidents?</h1>

In [33]:
accident['Month'] = pd.to_datetime(accident['Accident Date'], dayfirst=True).dt.month
accident['Month'].value_counts().sort_index()

Month
1     52872
2     49491
3     54086
4     51744
5     56352
6     56481
7     57445
8     53913
9     56455
10    59580
11    60424
12    51836
Name: count, dtype: int64

<h2>insigth: the month that has the highest number of accidents on the  record is november.</h2>

<h1> 7. which district area reports the most accidents overall?</h1>

In [34]:
accident['District Area'].value_counts()

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h2>insigth: the area that has the most accident is Birmingham.</h2>

<h1> 8. which day of the week has the highest accident frequency?</h1>

In [35]:
accident['dayofweek'].value_counts()

dayofweek
5    107178
2     99558
3     99511
4     97900
1     94550
6     89302
0     72680
Name: count, dtype: int64

<h2>insigth: the day of week that has he highest accident is 5 with the numbers of 107178.</h2>

<h1>9. which area is most likely to have fatal accident?</h1>

In [36]:
accident.groupby(['Urban_or_Rural_Area','Accident_Severity']).size()

Urban_or_Rural_Area  Accident_Severity
Rural                Fatal                  5601
                     Serious               37312
                     Slight               196077
Unallocated          Serious                   1
                     Slight                   10
Urban                Fatal                  3060
                     Serious               50904
                     Slight               367714
dtype: int64

<h2>insigth: Rural area is more likely to be prone in fatal accident.</h2>

<h1>10. do accidents occur more frequently on weekends or weekdays? </h1>

In [37]:
accident['dayofweek'] = pd.to_datetime(accident['Accident Date'], dayfirst=True).dt.day_name()
accident['dayofweek'].value_counts()

dayofweek
Saturday     107178
Wednesday     99558
Thursday      99511
Friday        97900
Tuesday       94550
Sunday        89302
Monday        72680
Name: count, dtype: int64

<h2>insight: it occurs mostly on weekends</h2>

<h1>11. where more accidents rural or urban?</h1>

In [38]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h2>insight: it seems thst the most accident record are lkely from urban area.</h2>

<h1>12. wich road condition  had the most serious accident?</h1>

In [39]:
accident[accident['Accident_Severity']=='Serious']['Road_Surface_Conditions'].value_counts()

Road_Surface_Conditions
Dry                     61708
Wet or damp             23785
Frost or ice             2007
Snow                      565
Flood over 3cm. deep      152
Name: count, dtype: int64

<h2>insight: the result shows that the most serious accident are likely at wet conditions</h2>

<h1>13. average casualties in rural areas?</h1>

In [40]:
accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

<h2>insight: based on the result above, it indicates that average casualties in rural areas has the record of 1.479204. </h2>

<h1>14. average casualties in urban areas?</h1>

In [41]:
accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

<h2>insight: based on the result above, it indicates that average casualties in rural areas has the record of 1.287807</h2>

<h1>15. district with the most accident during rain?</h1>

In [45]:
accident[accident['Weather_Conditions'].str.contains('Rain', na=False)]['District Area'].value_counts()

District Area
Birmingham                   1843
Glasgow City                 1139
Manchester                   1112
Leeds                         967
Sheffield                     876
                             ... 
Chester-le-Street              24
Teesdale                       21
London Airport (Heathrow)      20
Berwick-upon-Tweed             19
Clackmannanshire               11
Name: count, Length: 422, dtype: int64

<h2>insight: it shows that the district with the most accident during rain is birmingham</h2>

<h1>16. average number of vehicles per accident in rainy condition?</h1>

In [47]:
accident.groupby('Road_Surface_Conditions')['Number_of_Vehicles'].mean()

Road_Surface_Conditions
Dry                     1.841631
Flood over 3cm. deep    1.610619
Frost or ice            1.640006
Snow                    1.746859
Wet or damp             1.829161
Name: Number_of_Vehicles, dtype: float64

<h2>insight: the average number of vehicles per accident in rainy condition is 1.829161</h2>

<h1>17. average number of vehicles per accident in dry condition?</h1>

In [48]:
accident.groupby('Road_Surface_Conditions')['Number_of_Vehicles'].mean()

Road_Surface_Conditions
Dry                     1.841631
Flood over 3cm. deep    1.610619
Frost or ice            1.640006
Snow                    1.746859
Wet or damp             1.829161
Name: Number_of_Vehicles, dtype: float64

<h2>insight: it shows that the average condition per accident in dry condition is 1.841631</h2>

<h1>18. top 3 vehicle types in urban accident?</h1>

In [49]:
accident[accident['Urban_or_Rural_Area']=='Urban']['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      316062
Van / Goods 3.5 tonnes mgw or under       22003
Bus or coach (17 or more pass seats)      16851
Motorcycle over 500cc                     16700
Goods 7.5 tonnes mgw and over             11151
Motorcycle 125cc and under                10246
Taxi/Private hire car                      8617
Motorcycle over 125cc and up to 500cc      4982
Motorcycle 50cc and under                  4893
Goods over 3.5t. and under 7.5t            3864
Other vehicle                              3643
Agricultural vehicle                       1272
Minibus (8 - 16 passenger seats)           1258
Pedal cycle                                 127
Data missing or out of range                  6
Ridden horse                                  3
Name: count, dtype: int64

<h2>insight: based on the result above, it shows that the top 3 vehicles are car,van,and bus</h2>

<h1>19. least vehicle type in rural accident?</h1>

In [50]:
accident[accident['Urban_or_Rural_Area']=='Rural']['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      181922
Van / Goods 3.5 tonnes mgw or under       12156
Bus or coach (17 or more pass seats)       9025
Motorcycle over 500cc                      8957
Goods 7.5 tonnes mgw and over              6156
Motorcycle 125cc and under                 5023
Taxi/Private hire car                      4677
Motorcycle 50cc and under                  2710
Motorcycle over 125cc and up to 500cc      2674
Goods over 3.5t. and under 7.5t            2232
Other vehicle                              1994
Minibus (8 - 16 passenger seats)            718
Agricultural vehicle                        675
Pedal cycle                                  70
Ridden horse                                  1
Name: count, dtype: int64

<h2>insight: it shows that the least vehicle that involve in an accident is ridden horse</h2>

<h1>20. median vehicle involve in accidents?</h1>

In [51]:
accident['Number_of_Casualties'].mean()

1.357040257068864

<h2>insight: it says that the median vehicle involve in accident is 1.357040257068864</h2>

<h1>21. how many accidents accurred each year?</h1>

In [52]:
pd.to_datetime(accident['Accident Date'], dayfirst=True).dt.year.value_counts()

Accident Date
2019    182115
2020    170591
2021    163554
2022    144419
Name: count, dtype: int64

<h2>insight: it shows that the number of accident occurred each year decreases over time</h2>

<h1>22. top district with the highest number of accidents? </h1>

In [53]:
accident.groupby('District Area')['Number_of_Casualties'].sum()

District Area
Aberdeen City    1508
Aberdeenshire    2529
Adur              817
Allerdale        1663
Alnwick           351
                 ... 
Wychavon         1943
Wycombe          2442
Wyre             1706
Wyre Forest      1340
York             2453
Name: Number_of_Casualties, Length: 422, dtype: int64

<h2>insight: it shows that the top district with the highest number of casualties is Aberdeenshire</h2>

<h1>23. most common light condition during accident?</h1>

In [58]:
accident['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h2>insight: based on the result above, the most common light condition during accident is daylight</h2>

<h1>24. average casualties by weather condition?</h1>

In [63]:
accident.groupby('Weather_Conditions')['Number_of_Casualties'].mean()

Weather_Conditions
Fine + high winds        1.386018
Fine no high winds       1.347397
Fog or mist              1.452948
Other                    1.354869
Raining + high winds     1.416641
Raining no high winds    1.408214
Snowing + high winds     1.418079
Snowing no high winds    1.341776
Name: Number_of_Casualties, dtype: float64

<h2>insight: the average casualties by weather condition that is shows above is 1.386018</h2>

<h1>25. which light condition saw the most serious accident?</h1>

In [65]:
accident[accident['Accident_Severity']=='Serious']['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       60759
Darkness - lights lit          19130
Darkness - no lighting          7174
Darkness - lighting unknown      794
Darkness - lights unlit          360
Name: count, dtype: int64

<h2>insight: the ligh condition with the most serious accident is daylight</h2>